File: MPC.cpp
#include <iostream>
#include <chrono>
#include <vector>
#include <map>
#include <unistd.h>
#include "casadi/casadi.hpp"
using namespace casadi;
using namespace std;
#define Vmax 2.0

class MPC
{
    private:
        double T = 5.0;
        int N = 10;
        double dt = T / N;
        const int nx = 3;
        const int nu = 3;
        const vector<double> Q = {1.0, 1.0, 0.01}; //状態への重み
        const vector<double> R = {1.0, 1.0, 1.0};  //制御入力への重み

        vector<MX> w;        //最適化変数list
        vector<MX> w0;       //最適化変数wの初期推定解を格納
        vector<MX> lbw;      //最適化変数w lower bound
        vector<MX> ubw;      //最適化変数w upper bound
        MX J = MX::zeros(1); //コスト関数
        vector<MX> g;        //制約(等式制約,不等式制約)
        vector<MX> lbg;      //制約関数g lower bound
        vector<MX> ubg;      //制約関数g upper bound

        MX Xk = MX::sym("X0", nx); //初期時刻状態ベクトルx0
        vector<double> x_ref;
        MX Uk;
        int k_max = 0;
        
        MX solver(vector<double> inipos, vector<double> iniv, vector<vector<double>> path);
    public:
        MPC();
        std::map<std::string, DM> update(vector<double> inipos, vector<double> iniv, vector<vector<double>> path);
        vector<double> Tgrid();
};
MPC::MPC()
{
}
MX MPC::solver(vector<double> inipos, vector<double> iniv, vector<vector<double>> path)
{
    MX J = MX::zeros(1); //コスト関数
    for (int k = 0; k < this->N; k++)
    {
        this->Uk = MX::sym("U_" + to_string(k), this->nu);
        this->w.push_back(this->Uk);
        this->lbw.push_back(-Vmax);
        this->lbw.push_back(-Vmax);
        this->lbw.push_back(-Vmax);
        this->ubw.push_back(Vmax);
        this->ubw.push_back(Vmax);
        this->ubw.push_back(Vmax);
        this->w0.push_back(iniv[0]);
        this->w0.push_back(iniv[1]);
        this->w0.push_back(iniv[2]);
        //運動方程式
        MX x = this->Xk(0); // X座標[m]
        MX y = this->Xk(1); // Y座標[m]
        MX yaw = this->Xk(2); //ロボット角度[rad]
        //制御入力
        MX vx = this->Uk(0); // vx[m/s]
        MX vy = this->Uk(1); // vy[m/s]
        MX angular = this->Uk(2); // w[rad/s]
        this->k_max += 1;
        MX L = MX::zeros(1); //ステージコスト
        this->x_ref = path[k];
        cout << path[k];
        for (int i = 0; i < this->nx; i++)
            L += 0.5 * this->Q[i] * pow((this->Xk(i) - this->x_ref[i]), 2);
        for (int i = 0; i < this->nu; i++)
            L += 0.5 * this->R[i] * pow(this->Uk(i), 2);
        J = J + L * dt; //コスト関数にステージコストを追加
        // Forward Euler による離散化状態方程式
        MX Xk_next = vertcat(x + vx * dt, y + vy * dt, yaw + angular * dt);
        MX Xk1 = MX::sym("X_" + to_string(k + 1), this->nu); //時間ステージ k+1 の状態 xk+1 を表す変数
        this->w.push_back(Xk1);
        this->lbw.push_back(-inf);
        this->lbw.push_back(-inf);
        this->lbw.push_back(-inf);
        this->ubw.push_back(inf);
        this->ubw.push_back(inf);
        this->ubw.push_back(inf);
        this->w0.push_back(inipos[0]);
        this->w0.push_back(inipos[1]);
        this->w0.push_back(inipos[2]);
        //状態方程式(xk+1=xk+fk*dt) を等式制約として導入
        this->g.push_back(Xk_next - Xk1);
        this->lbg.push_back(0.0);
        this->lbg.push_back(0.0);
        this->lbg.push_back(0.0);
        this->ubg.push_back(0.0);
        this->ubg.push_back(0.0);
        this->ubg.push_back(0.0);
        this->Xk = Xk1;
    }
    MX Vf = MX::zeros(1); //終端コスト
    for (int i = 0; i < this->nx; i++)
        Vf += 0.5 * this->Q[i] * pow((this->Xk(i) - this->x_ref[i]), 2);
    for (int i = 0; i < this->nu; i++)
        Vf += 0.5 * this->R[i] * pow(this->Uk(i), 2);
    J = J + Vf;
    return J;
}
std::map<std::string, DM> MPC::update(vector<double> inipos, vector<double> iniv, vector<vector<double>> path)
{
    this->w.push_back(Xk);

    this->lbw.push_back(inipos[0]);
    this->lbw.push_back(inipos[1]);
    this->lbw.push_back(inipos[2]);

    this->ubw.push_back(inipos[0]);
    this->ubw.push_back(inipos[1]);
    this->ubw.push_back(inipos[2]);

    this->w0.push_back(0.0);
    this->w0.push_back(0.0);
    this->w0.push_back(0.0);
    MX J = solver(inipos, iniv,path);
    // NLP
    MXDict nlp = {{"f", J}, {"x", vertcat(this->w)}, {"g", vertcat(this->g)}};
    // Create NLP solver and buffers
    Function solver = nlpsol("solver", "ipopt", nlp);
    std::map<std::string, DM> arg, res;
    // Solve the NLP
    arg["x0"] = w0;
    arg["lbx"] = lbw;
    arg["ubx"] = ubw;
    arg["lbg"] = lbg;
    arg["ubg"] = ubg;
    res = solver(arg);
    return res;
}
vector<double> MPC::Tgrid()
{
    vector<double> tgrid;
    for (int k = 0; k < this->k_max + 1; k++)
        tgrid.push_back(this->dt * k);
    return tgrid;
}

vector<double> line_excerpt(vector<vector<double>> val, int line, int col = 0)
{
    vector<double> cut;
    if (col == 0)
        col = val.size();
    for (int i = 0; i < col; i++)
        cut.push_back(val[i][line]);
    return cut;
}

vector<double> slice(vector<double> val, int ini, int step)
{
    vector<double> cut;
    for (int i = ini; i < (int)val.size(); i += step)
        cut.push_back(val[i]);
    return cut;
}
#include "matplotlibcpp.h"
namespace plt = matplotlibcpp;
double Y = 0.0;
std::vector<std::vector<double>> path =
    {
        {1.0, Y, 0.0},
        {2.0, Y, 0.0},
        {3.0, Y, 0.0},
        {4.0, Y, 0.0},
        {5.0, Y, 0.0},
        {6.0, Y, 0.0},
        {7.0, Y, 0.0},
        {8.0, Y, 0.0},
        {9.0, Y, 0.0},
        {10.0, Y, 0.0},
        {11.0, Y, 0.0},
        {12.0, Y, 0.0},
        {13.0, Y, 0.0},
        {14.0, Y, 0.0},
        {15.0, Y, 0.0},
        {16.0, Y, 0.0},
        {17.0, Y, 0.0},
        {18.0, Y, 0.0},
        {19.0, Y, 0.0},
        {20.0, Y, 0.0},
        {21.0, Y, 0.0},
        {22.0, Y, 0.0},
        {23.0, Y, 0.0},
        {24.0, Y, 0.0},
        {25.0, Y, 0.0},
        {26.0, Y, 0.0},
        {27.0, Y, 0.0},
        {28.0, Y, 0.0},
        {29.0, Y, 0.0},
        {30.0, Y, 0.0},
        {31.0, Y, 0.0},
        {32.0, Y, 0.0},
        {33.0, Y, 0.0},
        {34.0, Y, 0.0},
        {35.0, Y, 0.0},
        {36.0, Y, 0.0},
        {37.0, Y, 0.0},
        {38.0, Y, 0.0},
        {39.0, Y, 0.0},
        {40.0, Y, 0.0},
        {41.0, Y, 0.0},
        {42.0, Y, 0.0},
        {43.0, Y, 0.0},
        {44.0, Y, 0.0},
        {45.0, Y, 0.0},
        {46.0, Y, 0.0},
        {47.0, Y, 0.0},
        {48.0, Y, 0.0},
        {49.0, Y, 0.0},
        {50.0, Y, 0.0},
        {49.0, Y, 0.0},
        {50.0, Y, 0.0},
        {51.0, Y, 0.0},
        {52.0, Y, 0.0},
        {53.0, Y, 0.0},
        {54.0, Y, 0.0},
        {55.0, Y, 0.0},
        {56.0, Y, 0.0},
        {57.0, Y, 0.0},
        {58.0, Y, 0.0},
        {59.0, Y, 0.0},
        {60.0, Y, 0.0},
        {71.0, Y, 0.0},
        {72.0, Y, 0.0},
        {73.0, Y, 0.0},
        {74.0, Y, 0.0},
        {75.0, Y, 0.0},
        {76.0, Y, 0.0},
        {77.0, Y, 0.0},
        {78.0, Y, 0.0},
        {79.0, Y, 0.0},
        {80.0, Y, 0.0},
        {81.0, Y, 0.0},
        {82.0, Y, 0.0},
        {83.0, Y, 0.0},
        {84.0, Y, 0.0},
        {85.0, Y, 0.0},
        {86.0, Y, 0.0},
        {87.0, Y, 0.0},
        {88.0, Y, 0.0},
        {89.0, Y, 0.0},
        {90.0, Y, 0.0},
        {91.0, Y, 0.0},
        {92.0, Y, 0.0},
        {93.0, Y, 0.0},
        {94.0, Y, 0.0},
        {95.0, Y, 0.0},
        {96.0, Y, 0.0},
        {97.0, Y, 0.0},
        {98.0, Y, 0.0},
        {99.0, Y, 0.0},
        {100.0, Y, 0.0},
        {101.0, Y, 0.0},
        {102.0, Y, 0.0}};

int main()
{
    MPC mpc;
    std::vector<double> inipos = {0.0, 1.0, 0.0};
    std::vector<double> iniv = {0.0, 0.0, 0.0};

    std::map<std::string, DM> res = mpc.update(inipos, iniv, path);
    auto x1_opt = slice((vector<double>)res.at("x"), 0, 6);
    auto x2_opt = slice((vector<double>)res.at("x"), 1, 6);
    auto x3_opt = slice((vector<double>)res.at("x"), 2, 6);
    auto u1_opt = slice((vector<double>)res.at("x"), 3, 6);
    auto u2_opt = slice((vector<double>)res.at("x"), 4, 6);
    auto u3_opt = slice((vector<double>)res.at("x"), 5, 6);
    // cout << x1_opt << endl;
    // cout << x2_opt << endl;
    // cout << x3_opt << endl;
    // cout << u1_opt << endl;
    // cout << u2_opt << endl;
    // cout << u3_opt << endl;
    vector<double> tgrid = mpc.Tgrid();
    plt::subplot(1, 2, 1);
    plt::grid(true);
    plt::plot(tgrid, x1_opt, "-b");
    plt::plot(tgrid, x2_opt, "-g");
    plt::plot(tgrid, x3_opt, "-r");

    plt::plot(tgrid, (vector<double>)vertcat(DM::nan(1), u1_opt), "-.");
    plt::plot(tgrid, (vector<double>)vertcat(DM::nan(1), u2_opt), "-.");
    plt::plot(tgrid, (vector<double>)vertcat(DM::nan(1), u3_opt), "-.");
    plt::subplot(1, 2, 2);
    plt::grid(true);
    plt::plot(line_excerpt(path, 0), line_excerpt(path, 1), "-r");
    plt::plot(x1_opt, x2_opt, ".g");
    plt::show();
    return 0;
}

File: MPC/mpc.py
from casadi import *
import numpy as np
import math
import matplotlib.pyplot as plt

class MPC(object):
    # 問題設定
    T = 30.0     # ホライゾン長さ
    N = 10     # ホライゾン離散化グリッド数
    dt = T / N  # 離散化ステップ
    nx = 3      # 状態空間の次元
    nu = 3      # 制御入力の次元
    Vmax = 1.0
    Angularmax = 1.0
    #ステージコスト
    Q = [100.0, 100.0, 50.0] # 状態への重み
    R = [1.0, 0.0, 1.0]  # 制御入力への重み
    #終端コスト
    Qf = [1.0, 1.0, 0.5] # 状態への重み
    Rf = [0.1, 0.0, 0.1]  # 制御入力への重み
    inipos=[]
    path=[]

    def __init__(self):
        pass
    def __init__(self,T,dt,Vmax,Angularmax):
        self.T=T
        self.dt=dt
        self.N=int(T/dt)
        self.Vmax=Vmax
        self.Angularmax=Angularmax

    def calc_vel(self,path):#inipos=[x0,y0,yaw0] path=[[x,y,yaw],...]
        cpath=[]
        size=len(path)
        acc = size/4
        v=0.0
        theta_pre=0.0
        for i in range(size):
            x = path[i][0]
            y = path[i][1]
            if i<(size-1):
                x_ = path[i+1][0]
                y_ = path[i+1][1]
            else:
                x_ = path[-1][0]
                y_ = path[-1][1]
            theta = math.atan2(y_-y,x_-x)
            diff_angle=theta-theta_pre
            theta_pre=theta
            #angular=diff_angle*self.dt
            angular=diff_angle
            if i<acc:
                v+=self.Vmax/acc
            elif i>(size-1)-acc:
                v-=self.Vmax/acc
            if v>self.Vmax:
                v=self.Vmax
            if v<0.0:
                v=0.0
            cpath.append([x,y,theta,v,0.0,angular])
            #print(cpath)
        return np.array(cpath)

    def bezier_curve(self,points, nTimes=1000):
        """
        Given a set of control points, return the
        bezier curve defined by the control points.
        points should be a list of lists, or list of tuples
        such as [ [1,1], 
                    [2,3], 
                    [4,5], ..[Xn, Yn] ]
            nTimes is the number of time steps, defaults to 1000
            See http://processingjs.nihongoresources.com/bezierinfo/
        """
        from scipy.special import comb
        def bernstein_poly(i, n, t):
            """
            The Bernstein polynomial of n, i as a function of t
            """
            return comb(n, i) * ( t**(n-i) ) * (1 - t)**i
        nPoints = len(points)
        xPoints = np.array([p[0] for p in points])
        yPoints = np.array([p[1] for p in points])

        t = np.linspace(0.0, 1.0, nTimes)

        polynomial_array = np.array([ bernstein_poly(i, nPoints-1, t) for i in range(0, nPoints)   ])

        xvals = np.dot(xPoints, polynomial_array)
        yvals = np.dot(yPoints, polynomial_array)

        return xvals, yvals

    def bezier_path(self,path_,n=1000):
        #path=np.vstack((init,path))#初期位置をpathに追加
        xvals, yvals = self.bezier_curve(path_, nTimes=n)#ベジェ曲線で経路を滑らかにする
        cpath=np.flipud(np.array(list(map(list, zip(xvals,yvals)))))#xvalsとyvalsの結合と反転
        #結果表示
        plt.plot(path_[:,0],path_[:,1])
        plt.plot(cpath[:,0],cpath[:,1],color = "red")
        plt.show()
        return cpath

    def conversion_path(self,path):
        l=0.0
        for i in range(len(path)-1):
            p=path[i]
            p_next=path[i+1]
            print(p,p_next)
            l+=math.sqrt((p[0]-p_next[0])**2+(p[1]-p_next[1])**2)
        n=int(l/self.dt)
        print("l:",l,"ntimes:",n)
        bpath=self.bezier_path(path,n)
        cpath=self.calc_vel(bpath)
        return cpath
        

    def cost(self,inipos,iniv,path):#inipos=[x0,y0,yaw0] path=[[x,y,yaw,vx,vy,angular],...]
        J = 0 # コスト関数
        for k in range(self.N):
            Uk = MX.sym('U_' + str(k), self.nu) # 時間ステージ k の制御入力 uk を表す変数
            self.w   += [Uk]                    # uk を最適化変数 list に追加
            #制御入力の制約条件
            self.lbw += [-self.Vmax,0.0,-self.Angularmax]        # uk の lower-bound
            self.ubw += [self.Vmax,0.0,self.Angularmax]           # uk の upper-bound

            self.w0  += iniv                 # uk の初期推定解

            #運動方程式
            x   = self.Xk[0]    # X座標[m]
            y   = self.Xk[1]    # Y座標[m]
            yaw = self.Xk[2]    # ロボット角度[rad]
            #制御入力
            vx      = Uk[0]   # vx[m/s]
            #vy      = Uk[1]   # vy[m/s]
            angular = Uk[2]   # w[rad/s]
            # ステージコストのパラメータ
            self.k_max+=1
            x_ref = path[k][0:3]           # 目標状態
            u_ref = path[k][3:6]           # 目標状態
            L = 0 # ステージコスト
            for i in range(self.nx):
                L += 0.5 * self.Q[i] * (self.Xk[i]-x_ref[i])**2
            for i in range(self.nu):
                L += 0.5 * self.R[i] * (Uk[i]-u_ref[i])**2
            J = J + L * self.dt # コスト関数にステージコストを追加

            # Forward Euler による離散化状態方程式
            Xk_next = vertcat(x + vx*cos(yaw)*self.dt,
                            y + vx*sin(yaw)*self.dt,
                            yaw + angular*self.dt)
            Xk1 = MX.sym('X_' + str(k+1), self.nx)  # 時間ステージ k+1 の状態 xk+1 を表す変数
            self.w   += [Xk1]                       # xk+1 を最適化変数 list に追加
            self.lbw += [-inf, -inf, -inf]    # xk+1 の lower-bound （指定しない要素は -inf）
            self.ubw += [inf, inf, inf]       # xk+1 の upper-bound （指定しない要素は inf）
            self.w0  += inipos       # xk+1 の初期推定解

            self.g   += [Xk_next-Xk1] # 状態方程式(xk+1=xk+fk*dt) を等式制約として導入
            self.lbg += [0.0,0.0,0.0] # 等式制約は lower-bound と upper-bound を同じ値にすることで設定
            self.ubg += [0.0,0.0,0.0] # 等式制約は lower-bo und と upper-bound を同じ値にすることで設定
            self.Xk = Xk1

        # 終端コストのパラメータ
        Vf = 0                            # 終端コスト
        for i in range(self.nx):
            Vf += 0.5 * self.Qf[i] * (self.Xk[i]-x_ref[i])**2
        for i in range(self.nu):
            Vf += 0.5 * self.Rf[i] * (Uk[i]-u_ref[i])**2
        J = J + Vf
        return J
    def solve(self,inipos,iniv,path):#inipos=[x0,y0,yaw0] path=[[x,y,yaw,vx,vy,angular],...]
        self.inipos=inipos
        self.path=path
        # 以下で非線形計画問題(NLP)を定式化
        self.w   = []  # 最適化変数を格納する list
        self.w0  = []  # 最適化変数(w)の初期推定解を格納する list
        self.lbw = []  # 最適化変数(w)の lower bound を格納する list
        self.ubw = []  # 最適化変数(w)の upper bound を格納する list
        self.g = []    # 制約（等式制約，不等式制約どちらも）を格納する list
        self.lbg = []  # 制約関数(g)の lower bound を格納する list
        self.ubg = []  # 制約関数(g)の upper bound を格納する list

        self.Xk = MX.sym('X0', self.nx) # 初期時刻の状態ベクトル x0

        self.w += [self.Xk]             # x0 を 最適化変数 list (w) に追加
        # 初期状態は given という条件を等式制約として考慮
        self.lbw += inipos # 等式制約は lower-bound と upper-bound を同じ値にすることで設定
        self.ubw += inipos # 等式制約は lower-bound と upper-bound を同じ値にすることで設定
        self.w0  += inipos # x0 の初期推定解
        self.k_max=0
        #現在位置から一番近い経路の点取得
        d=(path[:,0]-inipos[0])**2+(path[:,1]-inipos[1])**2
        min_d=np.argmin(d)
        path_=list(path[min_d:,:])#経路の現在位置から一番近い経路の点からゴールまでを抜き出し
        #サイズNのリストのあまりをゴールの座標で埋める
        for i in range((self.N-len(path[min_d:,:]))):
            path_.append(path[-1,:])
        path_=np.array(path_)
        #評価関数式作成
        J=self.cost(inipos,iniv,path_)
        # 非線形計画問題(NLP)
        nlp = {'f': J, 'x': vertcat(*self.w), 'g': vertcat(*self.g)}
        # Ipopt ソルバー，最小バリアパラメータを0.001に設定
        solver = nlpsol('solver', 'ipopt', nlp)

        # NLPを解く
        sol = solver(x0=self.w0, lbx=self.lbw, ubx=self.ubw, lbg=self.lbg, ubg=self.ubg)
        w_opt = sol['x'].full().flatten()
        print(self.k_max)
        tgrid = np.array([self.dt*k for k in range(self.k_max+1)])
        return w_opt, tgrid

    def goal_check(self,lim,alim):
        d=math.sqrt((self.path[-1,0]-self.inipos[0])**2+(self.path[-1,1]-self.inipos[1])**2)
        print(d,math.fabs(self.path[-1,2]-self.inipos[2]))
        if d<lim:
            if math.fabs(self.path[-1,2]-self.inipos[2])<alim:
                return True
        return False